### PR TITLE
Http avoid query parameters to be duplicated

### PR DESCRIPTION
**Pull request by vhvb1989**
Adding code to validate if a query parameter was previously added.
If yes, only the actual value is updated instead of inserting a new query parameter.

There is a little performance impact while adding a query parameter because I am doing a lineal search every time to check if the parameter is already there, and also when the value is updated to a different size than previous because it requires a memory shifting right or left on the url.
This is because we don't have something like a SET for the query parameters. We just use a pre-allocated buffer.

Also. The case of someone adding an initial URL with duplicated query parameters like `http://url.com?p1=1&p1=2` is not covered.

fixes: https://github.com/Azure/azure-sdk-for-c/issues/517

### CODE REVIEW

File: sdk/src/azure/core/az_http_request.c [hunk 1]
```diff
@@ -97,6 +101,66 @@ AZ_NODISCARD az_result az_http_request_append_path(_az_http_request* ref_request
   return AZ_OK;
 }
 
+// returns the query parameter value from a query paramenter name.
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:104]:
```suggestion
// returns the query parameter value from a query parameter name.
```

File: sdk/src/azure/core/az_http_request.c [hunk 2]
```diff
@@ -97,6 +101,66 @@ AZ_NODISCARD az_result az_http_request_append_path(_az_http_request* ref_request
...
+      is_query_start = 1; // set next index to be queryStart
+      if (value_index_start > 0)
+      { // if this was set before, it means we found the end of the value we want to return
+        return az_span_slice(url, value_index_start, index);
+      }
+      continue;
+    }
+    if (is_query_start == 1 && query_params_ptr[index] == new_qp_name_ptr[0])
+    { // at this point, a qp name start with the same letter as the new qp
+      // Check if they have same size before comparing content by cheking if
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:137]:
```suggestion
      // Check if they have same size before comparing content by checking if
```

File: sdk/src/azure/core/az_http_request.c [hunk 3]
```diff
@@ -107,11 +171,45 @@ az_http_request_set_query_parameter(_az_http_request* ref_request, az_span name,
...
+    AZ_RETURN_IF_NOT_ENOUGH_SIZE(url_remainder, required_length);
+
+    // There is already a query parameter name. update the value if different
+    if (az_span_is_content_equal(value, prexisting_query_parameter_value))
+    {
+      // No need to do anything, existing value is equal to the new one
+      return AZ_OK;
+    }
+
+    uint8_t* url_prt_start = az_span_ptr(ref_request->_internal.url);
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:195]:
```suggestion
    uint8_t* url_ptr_start = az_span_ptr(ref_request->_internal.url);
```

File: sdk/src/azure/core/az_http_request.c [hunk 4]
```diff
@@ -107,11 +171,45 @@ az_http_request_set_query_parameter(_az_http_request* ref_request, az_span name,
   // name or value can't be empty
   _az_PRECONDITION(az_span_size(name) > 0 && az_span_size(value) > 0);
 
-  int32_t required_length = az_span_size(name) + az_span_size(value) + 2;
-
+  // check if query parameter is already in url
+  az_span prexisting_query_parameter_value
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:175]:
```suggestion
  az_span pre_existing_query_parameter_value
```

File: sdk/src/azure/core/az_http_request.c [hunk 5]
```diff
@@ -97,6 +101,66 @@ AZ_NODISCARD az_result az_http_request_append_path(_az_http_request* ref_request
...
+    is_query_start = 0;
+  }
+
+  if (value_index_start > 0)
+  { // getting here means we found the qp at the last position of url, so the value goes to the end
+    // of url
+    return az_span_slice(url, value_index_start, ref_request->_internal.url_length);
+  }
+
+  // didn't find the query paramenter, return null
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:160]:
```suggestion
  // didn't find the query parameter, return null
```

File: sdk/src/azure/core/az_http_request.c [hunk 6]
```diff
@@ -97,6 +101,66 @@ AZ_NODISCARD az_result az_http_request_append_path(_az_http_request* ref_request
   return AZ_OK;
 }
 
+// returns the query parameter value from a query paramenter name.
+// Or returns AZ_SPAN_NULL if query parameter name is not found
+static AZ_NODISCARD az_span
+_az_http_request_find_query_parameter(_az_http_request* ref_request, az_span query_paramenter_name)
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:107]:
```suggestion
_az_http_request_find_query_parameter(_az_http_request* ref_request, az_span query_parameter_name)
```

File: sdk/src/azure/core/az_http_request.c [hunk 7]
```diff
@@ -97,6 +101,66 @@ AZ_NODISCARD az_result az_http_request_append_path(_az_http_request* ref_request
...
+  uint8_t* query_params_ptr = az_span_ptr(url);
+  uint8_t* new_qp_name_ptr = az_span_ptr(query_paramenter_name);
+  int32_t new_qp_name_size = az_span_size(query_paramenter_name);
+
+  // will be set to the start of the value of a qp equal to the new qp
+  int32_t value_index_start = -1;
+
+  for (int32_t is_query_start = 1, index = ref_request->_internal.query_start; // +1 to jump `?`
+       index < ref_request->_internal.url_length - new_qp_name_size;
+       index++)
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:124]: With the two loop-index variables, it becomes difficult to understand the logic of this loop. Can we extract out `is_query_start` into a local outside the loop?

File: sdk/src/azure/core/az_http_request.c [hunk 8]
```diff
@@ -97,6 +101,66 @@ AZ_NODISCARD az_result az_http_request_append_path(_az_http_request* ref_request
...
+        }
+      }
+    }
+    is_query_start = 0;
+  }
+
+  if (value_index_start > 0)
+  { // getting here means we found the qp at the last position of url, so the value goes to the end
+    // of url
+    return az_span_slice(url, value_index_start, ref_request->_internal.url_length);
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:157]: nit:
```suggestion
    return az_span_slice_to_end(url, value_index_start);
```

File: sdk/src/azure/core/az_http_request.c [hunk 9]
```diff
@@ -107,11 +171,45 @@ az_http_request_set_query_parameter(_az_http_request* ref_request, az_span name,
...
 
-  int32_t required_length = az_span_size(name) + az_span_size(value) + 2;
-
+  // check if query parameter is already in url
+  az_span prexisting_query_parameter_value
+      = _az_http_request_find_query_parameter(ref_request, name);
   az_span url_remainder
       = az_span_slice_to_end(ref_request->_internal.url, ref_request->_internal.url_length);
 
+  if (az_span_ptr(prexisting_query_parameter_value) != NULL)
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:180]: nit: Instead of checking the ptr is null, check that size == 0. Reduces the calls to az_span_ptr that we'd need to audit as code reviewers to make sure we aren't going out-of-bounds of the ptr.
```suggestion
  if (az_span_sizeprexisting_query_parameter_value) != 0)
```

File: sdk/src/azure/core/az_http_request.c [hunk 10]
```diff
@@ -107,11 +171,45 @@ az_http_request_set_query_parameter(_az_http_request* ref_request, az_span name,
...
+
+    // Replace the value content. This might shift right or left the url contents.
+    // No need to check
+    AZ_RETURN_IF_FAILED(_az_span_replace(
+        ref_request->_internal.url,
+        ref_request->_internal.url_length,
+        value_index,
+        value_index + prexisting_query_parameter_value_size,
+        value));
+    ref_request->_internal.url_length += difference;
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:207]: `difference` could be negative. Are we intentionally decreasing the `url_length` here by subtrating?

File: sdk/src/azure/core/az_http_request.c [hunk 11]
```diff
@@ -107,11 +171,45 @@ az_http_request_set_query_parameter(_az_http_request* ref_request, az_span name,
...
+    return AZ_OK;
+  }
+
+  // Adding new query parameter
+  int32_t required_length = az_span_size(name) + az_span_size(value) + 2;
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:212]: What does the `+ 2` mean here?

File: sdk/tests/core/test_az_http.c [hunk 12]
```diff
@@ -180,6 +180,222 @@ static void test_http_request(void** state)
...
+  { // Test adding duplicated query parameters
+    uint8_t buf[100];
+    uint8_t header_buf[(2 * sizeof(az_pair))];
+    memset(buf, 0, sizeof(buf));
+    memset(header_buf, 0, sizeof(header_buf));
+
+    az_span url_span = AZ_SPAN_FROM_BUFFER(buf);
+    az_span initial_url = AZ_SPAN_FROM_STR("http://example.com");
+    az_span remainder = az_span_copy(url_span, initial_url);
+    assert_int_equal(az_span_size(remainder), 100 - az_span_size(initial_url));
```
Comment (ahsonkhan) [inline sdk/tests/core/test_az_http.c:192]: nit: This assert isn't really necessary. It isn't testing anything related to http, just that the span copy worked.

File: sdk/tests/core/test_az_http.c [hunk 13]
```diff
@@ -180,6 +180,222 @@ static void test_http_request(void** state)
...
+    // set qp
+    assert_return_code(
+        az_http_request_set_query_parameter(
+            &request, AZ_SPAN_FROM_STR("q1"), AZ_SPAN_FROM_STR("v1")),
+        AZ_OK);
+
+    // set same qp new value
+    assert_return_code(
+        az_http_request_set_query_parameter(
+            &request, AZ_SPAN_FROM_STR("q1"), AZ_SPAN_FROM_STR("value1")),
```
Comment (ahsonkhan) [inline sdk/tests/core/test_az_http.c:214]: Are trailing spaces (or even embedded white-space characters) allowed within query parameters?

File: sdk/tests/core/test_az_http.c [hunk 14]
```diff
@@ -180,6 +180,222 @@ static void test_http_request(void** state)
...
+        az_span_size(initial_url),
+        header_span,
+        AZ_SPAN_FROM_STR("body")));
+
+    // set qp
+    assert_return_code(
+        az_http_request_set_query_parameter(
+            &request, AZ_SPAN_FROM_STR("q1"), AZ_SPAN_FROM_STR("v1")),
+        AZ_OK);
+
```
Comment (ahsonkhan) [inline sdk/tests/core/test_az_http.c:210]: nit: Add an assert that it was actually set to `v1`, before verifying it updated to `value1 `.

File: sdk/tests/core/test_az_http.c [hunk 15]
```diff
@@ -180,6 +180,222 @@ static void test_http_request(void** state)
     assert_return_code(
         az_http_request_append_header(&request, AZ_SPAN_FROM_STR("header"), AZ_SPAN_NULL), AZ_OK);
   }
+  { // Test adding duplicated query parameters
```
Comment (ahsonkhan) [inline sdk/tests/core/test_az_http.c:183]: Nice tests!

**Comment by ahsonkhan** [comment]
> Also. The case of someone adding an initial URL with duplicated query parameters like `http://url.com?p1=1&p1=2` is not covered.

What do you mean by that. Are you planning on adding tests/validation for that in a separate PR? What happens if someone was to do that as the initial url and then try to set `p1`?

File: sdk/src/azure/core/az_http_request.c [hunk 16]
```diff
@@ -107,11 +171,45 @@ az_http_request_set_query_parameter(_az_http_request* ref_request, az_span name,
...
+    // There is already a query parameter name. update the value if different
+    if (az_span_is_content_equal(value, prexisting_query_parameter_value))
+    {
+      // No need to do anything, existing value is equal to the new one
+      return AZ_OK;
+    }
+
+    uint8_t* url_prt_start = az_span_ptr(ref_request->_internal.url);
+    // ptr difference should work because value is always inside url, same memory
+    int32_t value_index = (int32_t)(az_span_ptr(prexisting_query_parameter_value) - url_prt_start);
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:197]: nit: use the existing `_az_span_diff(...)` helper method to reduce duplication/calls to `az_span_ptr`).

File: sdk/src/azure/core/az_http_request.c (hunk 7 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:124]: it is only used inside the loop. I am trying to follow a @JeffreyRichter  advise about avoiding to declare variables outside a loop if only required in loop. Then we know for sure that variable is never required again for anything

File: sdk/src/azure/core/az_http_request.c (hunk 8 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:157]: yeah, that was my first approach but it is actually wrong..  the `url_length` can be smaller than the url span size.
Imaging az_span url of size 100 with an url like `http://r.com?q=value`,  url_leght is 20 while az_span url size is 100, 
So az_span_slice_to_end will get `value` with size greater than 5 which is not correct

File: sdk/src/azure/core/az_http_request.c (hunk 9 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:180]: can't.  a query parameter can be 0 length.  Like `?query=&query2=&query3=value`
In that case, the az_span value we return is length 0 but with a valid ptr.

The only way we know they query was not found is with ptr=NULL

File: sdk/src/azure/core/az_http_request.c (hunk 10 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:207]: yeah, when difference is negative it means we are reducing the length of url like
query = `?q1=value1`
then we call set_query(q1, v1);
this makes query = `?q1=v1`,, with a negative difference that indicates how much the url length is decremented

File: sdk/src/azure/core/az_http_request.c (hunk 11 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:212]: required symbols to be added `&/?` and `=`   .  I believed there was a comment about it over there.

File: sdk/tests/core/test_az_http.c (hunk 12 above)
Comment (vhvb1989) [reply sdk/tests/core/test_az_http.c:192]: yeah,, it came together when copy pasting an existing test and just wanted to leave it there :)

File: sdk/tests/core/test_az_http.c (hunk 13 above)
Comment (vhvb1989) [reply sdk/tests/core/test_az_http.c:214]: yeah, but they need to be url-encoded.
I think we decided to remove url-encoding some time ago and expect customer to set/add query parameters already encoded.
White spaces become something like `%20%` in the query parameter value.  But we don't have that

File: sdk/src/azure/core/az_http_request.c (hunk 16 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:197]: I was forgetting about that API!  thank you!

**Comment by vhvb1989** [comment]
> 
> 
> > Also. The case of someone adding an initial URL with duplicated query parameters like `http://url.com?p1=1&p1=2` is not covered.
> 
> What do you mean by that. Are you planning on adding tests/validation for that in a separate PR? What happens if someone was to do that as the initial url and then try to set `p1`?

Not adding validation for that at all.  It means that if a customer decide to initially add one url with duplicated queryParameters, we won't do anything to prevent it.
We could either fail the request_init maybe, or we could take just the first/last query parameter as the only one. Or, like now, we just don't care about this case.

We can talk more about this with @JeffreyRichter to see if we want to sacrifice a little more performance to validate this upon request_init. But I think we should not.  At any case, it would be Azure Service decision how to handle an intentionally duplicated  query parameter by user.

File: sdk/src/azure/core/az_http_request.c (hunk 16 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:197]: just realized this API does not work here.
The API returns the diff after slicing one az_span B out of another az_span A, by doing sizeA - sizeB

What I need is just the distance from the start A to the start B. It works if B ends at same position as B, but if B ends before, this distance is not the one I want because I would get  also any distance after B

File: sdk/src/azure/core/az_http_request.c (hunk 9 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_http_request.c]: Ah, I see. Do you think it is feasible for us to change `_az_http_request_find_query_parameter` to return `az_result` and `out az_span` and maybe have the `az_result` indicate whether we found it or not (if we already have an enum value that would work in this case)?

That will help reduce dependency on checks like these which differentiate between empty span and a span with null ptr.

File: sdk/src/azure/core/az_http_request.c (hunk 16 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_http_request.c]: Yes, you are right. That method returns the difference in size, not the ptr diff. I was misremembering and the overlaps helper just returns a bool, not the integer difference.

**Comment by ahsonkhan** [comment]
> At any case, it would be Azure Service decision how to handle an intentionally duplicated query parameter by user.

That's fair, but I think we should still have some deterministic behavior in that edge case, even if we don't validate/reject. If we are going to always override the **first** one we find when we set the parameter after initializing, then we should add a test to make sure that remains true. Same if we decide that we update the last one found.

It would be odd, if today we replace the first one, and then tomorrow (or for some other request), we replace the last one instead.

File: sdk/tests/core/test_az_http.c (hunk 13 above)
Comment (ahsonkhan) [reply sdk/tests/core/test_az_http.c:222]: Gotcha. Either we url-encode them, or we  should reject them (possibly in a precondition). Letting it go through un-encoded could be problematic.

File: sdk/src/azure/core/az_http_request.c (hunk 7 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_http_request.c:126]: > advise about avoiding to declare variables outside a loop if only required in loop

I think that is a good goal, but in this case, the variable essentially is outside the scope of the loop. We are just intentionally defining it in the for-loop definition, to try to force into that pattern, causing a bit of confusion.

If you re-write this using a while loop, you wouldn't be able to put `int32_t is_query_start` inside the loop body.
Either way, just a nit.

File: sdk/src/azure/core/az_http_request.c [hunk 17]
```diff
@@ -107,11 +173,54 @@ az_http_request_set_query_parameter(_az_http_request* ref_request, az_span name,
...
+    ref_request->_internal.url_length += difference;
+    if (difference < 0)
+    {
+      // clear anything after a left shift
+      difference *= -1;
+      az_span_fill(
+          az_span_slice(
+              az_span_slice_to_end(ref_request->_internal.url, ref_request->_internal.url_length),
+              0,
+              difference),
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_request.c:216]: Can we clear everything beyond the end of `url_length`? Why do we slice from `0 to difference`?

File: sdk/tests/core/test_az_http.c (hunk 13 above)
Comment (vhvb1989) [reply sdk/tests/core/test_az_http.c:222]: I am considering that blocking this characters is going to be kind of expensive. We would need to search the initial URL when init request happen, and fail the init if there is an invalid char.
I am thinking we could document that query parameters should be encoded before adding. At the end, it is Azure Service Clients the ones who creates a Request (not final customers. So, documenting this could be enough.
Will check with @JeffreyRichter

File: sdk/src/azure/core/az_http_request.c (hunk 17 above)
Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:216]: performance. 
If we have a very huge url buffer, and shift left just one position multiple times, we would be filling and filling all the remaining bytes over and over.
That's why we just care about the content that was left shifted

Comment (ahsonkhan) [reply sdk/src/azure/core/az_http_request.c:216]: I see. In that case, is it possible for us to be in a state where query parameters/data from older requests (when using a shared array) persists? Or, since every set and shift left will clear the "remaining", will we never have old data in the buffer, beyond the url length?

Imagine you have this in your buffer, as the first request:
`?q1=123&q2=456`

Then you create a second request:
`?q1=123` (&q2=456 is still there in the buffer)
You set the `q1` to a smaller length value, causing a left shift.
`?q1=1\0\0` (&q2=456 is still there in the buffer)
We cleared the bytes `23`, but what about the data from the first request? That is still visible. I can see the value of `q2`.

Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:216]: yeah, I considered that one. But I am considering that all this method should care is about the leaking it might cause only.
The case you shared is not really this method fault right?
Don't we have the same case by re-using the shared array without even calling this method?  If an initial URL has queryparameters, and we never call set_query_(...), then we use the same buffer for a new request, we would see all query data in the url remainder span... And that's not `az_http_request_set_query_parameter` fault.

So, if we want to handle a case like that, we should not use `az_http_request_set_query_parameter` to clean, since it might be never called.
We would need to come up with another strategy, like using retry policy to clean query parameters upon completing a request, or something like that.

Also, consider that Requests are not exposed to public.  Request is internal and only used by SDK clients.  Each SDK client uses stack for the Request (url and headers). So, probably we should care more about the Response.

File: sdk/src/azure/core/az_http_request.c (hunk 11 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_http_request.c:223]: Where? I don't see one in the `az_http_request_set_query_parameter` method.

Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:223]: I can add it. No problem

Comment (vhvb1989) [reply sdk/src/azure/core/az_http_request.c:224]: Added
